### PR TITLE
fix: dim = 2 in fit_transform overwrites self.dim set in constructor

### PR DESCRIPTION
**Pull request by code-lukas**
Dear authors,

I noticed that there is an error in your implementation:

```python3
from hnne import HNNE
import numpy as np

data = np.random.random(size=(1000, 256))

hnne = HNNE(dim=3)
projection = hnne.fit_transform(data)

print(projection.shape)
>>> (1000, 2)
```
This appears to happen because `HNNE.fit_transform()` accepts an argument `dim` which is set to `2` by default.
Consequently, `self.dim` is always set to `2` if you do not pass an explicit argument to the method. I think this leads to a lot of confusion, so I fixed this by removing this logic altogether.

Testing the above example after the proposed changes yields the correct size.

Cheers!

### CODE REVIEW

**Comment by koulakis** [comment]
Hi Lucas, thanks for catching this bug!  Sorry for my super late reply :sweat_smile: 

Actually there is a reason for allowing the user to override this dimension, namely that once you fit hnne, you can then easily project new data to multiple dimensions with a very low computational cost. There are some use cases where one would like to do this. You can see an example usage in this notebook: https://github.com/koulakis/h-nne/blob/main/notebooks/demo2_multiple_projections.ipynb

It is unfortunately unclear from the function call what this dim should do. It is also debatable whether it belongs to the `fit` method as it does only part of the fit. On the other hand it does not belong to `transform` as transform expects a dimension to have been picked and the projection mapping to exist for it. 

Maybe one choice would be to fix the bug by setting the default to `dim: Optional[int] = None` and also rename `dim` to something more informative like `override_dim`. Then one could additionally extend a bit the description of this input in the docstring to explain that this should only be used in this specific use case. 

What do you think?

**Comment by code-lukas** [comment]
Hi Marios,

no worries, thanks for getting back to me!
I appreciate that you guys made such a neatly packaged module which is pretty much plug-and-play.

I very much like your commitment to adhere to the `sklearn`-style of using it. Other implementations (such as PCA, MDS, Isomap, etc.) call `dim` `n_components`, although I don't know if this is really a better name.  Still, most people using HNNE are probably familiar with this notation.

I like the changes you proposed, calling it `override_dim` or `override_n_components` already takes away a lot of the confusion.

These changes are fairly minor, but if you want me to assist, let me know.

Cheers!

**Comment by koulakis** [comment]
Hi Lucas,

Indeed `n_components` is adheres better to the `sklearn` style but let's keep `dim` for now as it would be a bigger change (one should change it in the class initialization and in the local variables, that would need quite some testing afterwards). If you want to contribute to the project (also given you spotted the bug), please go on and make the changes. If you are too busy, I could do them as well, whatever you prefer.

If you do so, please don't completely remove the old `dim` argument, but instead deprecate it. Maybe set its default value to 'deprecated' and raise a warning which mentions that it will be removed in future versions and `override_dim` should be used instead. Also we should adjust the name in the example notebook (I don't think we need to rerun it, just rename it and keep the cell outputs).

Best,
Marios
